source/include/microapp/cs_Microapp.h

```diff
@@ -1,99 +1,115 @@
-		void operator=(Microapp const &);
-
-		/**
-		 * The last message we have sent. It has a repeat value that will start at "number_of_notifications". If it
-		 * is at zero, it will not be repeated anymore.
-		 */
-		TYPIFY(EVT_MICROAPP) _prevMessage;
-
-		/**
-		 * For the last chunk we have to do some other things. Validate the entire app, etc.
-		 */
-		bool _lastChunk;
-
-		/**
-		 * Local flag to check if microapp is enabled.
-		 */
-		bool _enabled;
-
-		/**
-		 * Local flag to indicate that ram section has been loaded.
-		 */
-		bool _loaded;
-
-		/**
-		 * Debug mode
-		 */
-		bool _debug;
-
-	protected:
-		/**
-		 * Set IPC ram data.
-		 */
-		void setIpcRam();
-
-		/**
-		 * Enable app.
-		 */
-		uint16_t enableApp(bool flag);
-
-		/**
-		 * Check if app is enabled.
-		 */
-		bool isEnabled();
-
-		/**
-		 * Call the loop function
-		 */
-		void callLoop(int & cntr, int & skip);
-
-		/**
-		 * Load ram information, set by microapp.
-		 */
-		uint16_t interpretRamdata();
-	public:
-		static Microapp& getInstance() {
-			static Microapp instance;
-			return instance;
-		}
-
-		/**
-		 * Initialize fstorage. Allocate buffer.
-		 */
-		uint16_t init();
-
-		/**
-		 * The tick function is used to send more notifications than one.
-		 */
-		void tick();
-
-		/**
-		 * Handle incoming events.
-		 */
-		void handleEvent(event_t & event);
-
-		/**
-		 * Handle incoming packet. Returns err_code.
-		 */
-		uint32_t handlePacket(microapp_packet_header_t *packet_stub);
+public:
+	static Microapp& getInstance() {
+		static Microapp instance;
+		return instance;
+	}
+
+	/**
+	 * Initialize storage, and load microapps.
+	 */
+	void init();
+
+private:
+	/**
+	 * Singleton, constructor, also copy constructor, is private.
+	 */
+	Microapp();
+	Microapp(Microapp const&);
+	void operator=(Microapp const &);
+
+	/**
+	 * The state of each microapp.
+	 */
+	microapp_state_t _states[MAX_MICROAPPS];
+
+	/**
+	 * Local flag to indicate that ram section has been loaded.
+	 */
+	bool _loaded = false;
+
+	void loadApps();
+
+	void loadState(uint8_t index);
+
+	/**
+	 * Validates app: compare checksum.
+	 * App state is updated in this call, make sure to store the state afterwards.
+	 */
+	cs_ret_code_t validateApp(uint8_t index);
+
+	/**
+	 * Enables app: checks sdk version.
+	 * App state is updated in this call, make sure to store the state afterwards.
+	 */
+	cs_ret_code_t enableApp(uint8_t index);
+
+	/**
+	 * Start app, if it passed all tests.
+	 */
+	cs_ret_code_t startApp(uint8_t index);
+
+	/**
+	 * Resets app state in ram only.
+	 */
+	void resetState(uint8_t index);
+
+	/**
+	 * Resets app state tests in ram only.
+	 */
+	void resetTestState(uint8_t index);
+
+	/**
+	 * Store app state to flash.
+	 */
+	cs_ret_code_t storeState(uint8_t index);
+
+	/**
+	 * Checks app state and returns true if this app is allowed to run.
+	 */
+	bool canRunApp(uint8_t index);
+
+	/**
+	 * The tick function is used to send more notifications than one.
```

Comment:
Not sure if this is true, my interpretation would be:
```suggestion
	 * Processes pending notifications.
```

Reply:
Ah, that's old. Changed it to "To be called every tick. Calls all microapps."